zello-sdk/src/com/zello/sdk/headset/HeadsetBroadcastReceiver.kt

```diff
@@ -0,0 +1,46 @@
+package com.zello.sdk.headset
+
+import android.content.Intent
+import android.view.KeyEvent
+import androidx.media.session.MediaButtonReceiver
+import com.zello.sdk.headset.HeadsetBroadcastReceiver.Companion.onKeyEvent
+import java.lang.ref.WeakReference
+
+/**
+ * Broadcast receiver that is used with the media session
+ * and allows to receive headset media key events while the app is running
+ * in the background.
+ * Instances of this class are created by Android framework and therefore
+ * can't receive any parameters. A static [onKeyEvent] member should be
+ * used to register an actual recipient of the key events.
+ */
+class HeadsetBroadcastReceiver : MediaButtonReceiver() {
+
+	companion object {
+
+		/**
+		 * Handle a media button intent.
+		 */
+		fun handleIntent(intent: Intent): Boolean {
+			if (intent.action != Intent.ACTION_MEDIA_BUTTON) return false
+			val event = intent.getParcelableExtra<KeyEvent>(Intent.EXTRA_KEY_EVENT) ?: return false
+			onKeyEvent?.invoke(event)
+			return true
+		}
+
+		/**
+		 * Register a recipient for the key events.
+		 */
+		var onKeyEvent: ((KeyEvent) -> Unit)?
+			get() {
+				return handler?.get()
+			}
+			set(value) {
+				handler = if (value == null) null else WeakReference(value)
```

Comment:
I'm a bit apprehensive about storing this as a `WeakReference`. I don't think it's intuitive to callers of this method that they would need to retain a reference to the original lambda in order to make sure the callback keeps working. In the usage within this PR, the lambda originates inside a static method in `Headset.kt`, so wouldn't it very likely be GC'd?

For the sake of leak prevention, would a `register` and `unregister` method be a viable alternative do you think? In my mind given the sensitivity of missed callbacks in this case, a leak risk is preferable to risking the callback get removed unexpectedly. Especially since the callback itself is likely expected to survive the whole lifecycle of the application anyway

Reply:
You are right. Also I totally over-engineered this as the object lifecycle is managed internally by the SDK. If the headset component is stopped once it's not needed anymore, the callback is cleared.